### ISSUE TITLE
Can you make a new progress bar for each epoch?

### ISSUE THREAD

**Issue by bjourne** [PyTorchLightning/pytorch-lightning#2189]
The progress bar is very slick but a big problem with it is that it overwrites itself. For example, if you are at epoch 10, you cannot see what the validation and training losses were for epoch 9. Could the progress bar perhaps be made to work more like in Keras so that you can see the losses of accuracies of previous epochs?

**Comment by awaelchli** [comment]
Yes, I was thinking of adding a flag to the progressbar callback that could enable that, but holding back on it because we were talking about switching to fastprogress.

If you need it asap, simply try to inherit from from ProgressBar callback, set the `leave = True` in the tqdm bar and pass your new callback to the trainer, like so: `Trainer(callbacks=[myprogressbar])`

**Comment by bjourne** [comment]
I tried this
```python
class MyProgressBar(ProgressBar):
    def init_validation_tqdm(self):
        bar = super().init_validation_tqdm()
        bar.leave = True
        return bar

    def init_train_tqdm(self):
        bar = super().init_train_tqdm()
        bar.leave = True
        return bar

    def init_test_tqdm(self):
        bar = super().init_test_tqdm()
        bar.leave = True
        return bar
```
It seem to make a newline for the validation progress bar. But I can't see the values for training and validation losses and accuracies. I don't really need a progress bar - just those for values for each epoch.

**Comment by awaelchli** [comment]
okay, then you could try to access `trainer.progress_bar_metrics` (dict) and print them yourself on epoch end maybe?

**Comment by bjourne** [comment]
I'm a PyTorchLightning newbie so I don't know how to do that but I'll try.

**Comment by shijianjian** [comment]
I think the easiest way is to:
```python
def on_epoch_start(self):
    print('\n')
```
It will end this line and start a new tqdm line.

**Comment by burntcobalt** [comment]
This is ugly, but will give the OP what they asked for:
`    def training_epoch_end(self, outputs):
        self.trainer.progress_bar_callback.main_progress_bar.write(
            f"Epoch {self.trainer.current_epoch} training loss={self.trainer.progress_bar_dict['loss']}")

    def validation_epoch_end(self, outputs):
        loss = torch.stack(outputs).mean()
        self.trainer.progress_bar_callback.main_progress_bar.write(
            f"Epoch {self.trainer.current_epoch} validation loss={loss.item()}")`

**Comment by dsantiago** [comment]
If you like Keras progress bar, can use this code (it's incomplete, neeeds to add validation and test hooks)

```python
class KerasProgressBar(pl.callbacks.progress.ProgressBarBase):

  def on_train_start(self, trainer, pl_module):
    super().on_train_start(trainer, pl_module)
    self.n_batches = trainer.num_training_batches
    pbar = tf.keras.utils.Progbar(self.n_batches)
    self.keras_bar = pbar

  def on_train_epoch_start(self, trainer, pl_module):
    super().on_train_epoch_start(trainer, pl_module)
    print(f'Epoch {trainer.current_epoch+1}/{trainer.max_epochs}')

  def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx):
    super().on_train_batch_end(trainer, pl_module, outputs, batch, batch_idx, dataloader_idx)
    output = [(k,  float(v)) for k, v in trainer.progress_bar_dict.items()]
    self.keras_bar.update(batch_idx + 1, values=output)
    time.sleep(0.2)

bar = KerasProgressBar()
...
trainer = pl.Trainer(max_epochs=10, gpus=1, logger=False, callbacks=[bar])
```

**Comment by tshu-w** [comment]
@awaelchli Hi, can you consider reopening this issue, absolutely needs an option.

**Comment by awaelchli** [comment]
Here is a simple progress bar for epochs:

```python
from pytorch_lightning.callbacks import ProgressBarBase


class EpochProgressBar(ProgressBarBase):

    def __init__(self):
        super().__init__()
        self.bar = None

    def on_train_start(self, trainer, pl_module):
        self.bar = tqdm(
            desc='Epoch',
            leave=False,
            dynamic_ncols=True,
            total=trainer.max_epochs,
        )

    def on_train_epoch_end(self, *args, **kwargs):
        self.bar.update(1)
```

Use it like this: 

`Trainer(callbacks=EpochProgressBar())`

**Comment by tshu-w** [comment]
@awaelchli To clarify, I think the idea is to have a separate progressbar for each epoch, rather than overlaying it on top of the original. Here is an example from allennlp.
<img width="2032" alt="Screen Shot 2021-05-19 at 10 07 41" src="https://user-images.githubusercontent.com/13161779/118746285-8b7aa380-b88a-11eb-8566-4c96cd257ff2.png">

However, I don't know how to implement it. ProgressBarBase example in the doc does the same thing, with the later epochs overwriting the output of the earlier epochs.
```python
class LitProgressBar(ProgressBarBase):

    def __init__(self):
        super().__init__()  # don't forget this :)
        self.enable = True

    def disable(self):
        self.enable = False

    def on_train_batch_end(self, trainer, pl_module, outputs):
        super().on_train_batch_end(trainer, pl_module, outputs)  # don't forget this :)
        percent = (self.train_batch_idx / self.total_train_batches) * 100
        sys.stdout.flush()
        sys.stdout.write(f'{percent:.01f} percent complete \r')

bar = LitProgressBar()
trainer = Trainer(callbacks=[bar])
```

**Comment by awaelchli** [comment]
```python
    class LitProgressBar(ProgressBar):

        def on_train_epoch_end(self, *args, **kwargs):
            super().on_train_epoch_end(*args, **kwargs)
            print()
```

**Comment by tshu-w** [comment]
@awaelchli Thank you for your patience in replying, I find this may be better for others with such needs
```python
class LitProgressBar(ProgressBar):
    def on_train_epoch_start(self, trainer, pl_module):
        if trainer.current_epoch:
            print()
        super().on_train_epoch_start(trainer, pl_module)
```

**Comment by AlexPasqua** [comment]
> If you like Keras progress bar, can use this code (it's incomplete, neeeds to add validation and test hooks).
> Obs: Need to have tensorflow also installed, or just copy the class code from source, so you don't need to install tensorflow:
> https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/utils/generic_utils.py#L815-L1026
> 
> ```python
> class KerasProgressBar(pl.callbacks.progress.ProgressBarBase):
> 
>   def on_train_start(self, trainer, pl_module):
>     super().on_train_start(trainer, pl_module)
>     pbar = tf.keras.utils.Progbar(trainer.num_training_batches)
>     self.keras_bar = pbar
> 
>   def on_train_epoch_start(self, trainer, pl_module):
>     super().on_train_epoch_start(trainer, pl_module)
>     print(f'Epoch {trainer.current_epoch+1}/{trainer.max_epochs}')
> 
>   def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx):
>     super().on_train_batch_end(trainer, pl_module, outputs, batch, batch_idx, dataloader_idx)
>     output = [(k,  float(v)) for k, v in trainer.progress_bar_dict.items()]
>     self.keras_bar.update(batch_idx + 1, values=output)
> 
> bar = KerasProgressBar()
> ...
> trainer = pl.Trainer(max_epochs=10, gpus=1, logger=False, callbacks=[bar])
> ```

What about validation and test hooks? I've tried to write them, but it does not work. I'm new to Pytorch and Pytorch Lightning

**Comment by dsantiago** [comment]
> What about validation and test hooks? I've tried to write them, but it does not work. I'm new to Pytorch and Pytorch Lightning

You will need to add the needed observers, like `on_eval...`, `on_test...`, etc...

**Comment by keunwoochoi** [comment]
+1

**Comment by vedal** [comment]
+1

**Comment by peterdudfield** [comment]
> I think the easiest way is to:
> 
> ```python
> def on_epoch_start(self):
>     print('\n')
> ```
> 
> It will end this line and start a new tqdm line.

Thanks this is perfect - thanks

**Comment by youurayy** [comment]
Isn't this something that should work out of the box?

What I want:
1. Track my training progress visually over every epoch (like in Keras) - not having this out of the box really shocked me coming from Keras. Tensorboard doesn't replace this. (Multiple concurrent/parallel Progress Bars provide a fraction of the utility, imho.)
2. Be able to use textual TQDM in Jupyter (I don't like the Ipywidgets one: doesn't follow dark scheme + doesn't have monospaced font + removed validation bars leave "</>" line (all in VSCode Jupyter), so 50 epochs == 50 "</>" lines.
3. In text mode in VSCode Jupyter (with Ipywidgets package removed), the Validation bar flashes crazily. I know this is probably an issue of VSCode, but at the same time Keras did not have any of these issues + was simpler + much more informative.

Lighting is an awesome project but I think the Progress Bar design should be reconsidered.

**Comment by youurayy** [comment]
In case it helps anyone, and to demonstrate what I meant, here's a sample Keras-like progress bar implementation:

```python
class SimpleProgressBar(pl.callbacks.ProgressBar):
    def __init__(self):
        super().__init__()
        self.bar = None

    def on_train_epoch_start(self, trainer, pl_module):
        self.bar = tqdm(total=self.total_train_batches,
                        desc=f"Epoch {trainer.current_epoch+1}",
                        position=0,
                        leave=True)
        self.running_loss = 0.0

    def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx):
        self.running_loss += outputs['loss'].item()
        self.bar.update(1)
        self.bar.set_postfix(loss=self.running_loss / self.total_train_batches)
        # self.bar.set_postfix(self.get_metrics(trainer, pl_module))

    def on_train_epoch_end(self, trainer, pl_module):
        self.bar.close()

    def on_validation_epoch_end(self, trainer, pl_module) -> None:
        if self.bar:
            val_loss = trainer.logged_metrics['val_loss_epoch'].item()
            loss = self.running_loss / self.total_train_batches
            self.bar.set_postfix(loss=loss, val_loss=val_loss)
```

![image](https://user-images.githubusercontent.com/588573/232329582-44e9873b-e5a5-469b-ab3d-d27a1491e606.png)

**Comment by youurayy** [comment]
In case it helps anyone, and to demonstrate what I meant, here's a sample Keras-like progress bar implementation:

```python
from tqdm import tqdm
import lightning.pytorch as pl

class SimpleProgressBar(pl.callbacks.ProgressBar):
    def __init__(self):
        super().__init__()
        self.bar = None

    def on_train_epoch_start(self, trainer, pl_module):
        self.bar = tqdm(total=self.total_train_batches,
                        desc=f"Epoch {trainer.current_epoch+1}",
                        position=0,
                        leave=True)
        self.running_loss = 0.0

    def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx):
        self.running_loss += outputs['loss'].item()
        self.bar.update(1)
        loss = self.running_loss / self.total_train_batches
        self.bar.set_postfix(loss=f'{loss:.3f}')
        # self.bar.set_postfix(self.get_metrics(trainer, pl_module))

    def on_train_epoch_end(self, trainer, pl_module):
        self.bar.close()

    def on_validation_epoch_end(self, trainer, pl_module) -> None:
        if self.bar:
            val_loss = trainer.logged_metrics['val_loss_epoch'].item()
            loss = self.running_loss / self.total_train_batches
            self.bar.set_postfix(loss=f'{loss:.3f}', val_loss=f'{val_loss:.3f}')
```

![image](https://user-images.githubusercontent.com/588573/232330409-20a5ff3c-1535-4a23-9fdd-756e5098d58d.png)

**Comment by marcinwrochna** [comment]
The built-in RichProgressBar (using `rich` instead of `tqdm`) is available since at least v1.5.10 (2022) and has a `leave` argument:
```python
from pytorch_lightning.callbacks import RichProgressBar
trainer = pl.Trainer(callbacks=RichProgressBar(leave=True))
```

![Screenshot_20240128_113804](https://github.com/Lightning-AI/pytorch-lightning/assets/11655579/ab6680b6-855d-494d-85d2-90dee594ac32)

TQDMProgressBar could support that too.

**Comment by thunderock** [comment]
> ```python
> class LitProgressBar(ProgressBar):
>     def on_train_epoch_start(self, trainer, pl_module):
>         if trainer.current_epoch:
>             print()
>         super().on_train_epoch_start(trainer, pl_module)
> ```

Well, this doesn't work as expected when you have validation after every epoch.
